src/main/scala/org/ergoplatform/nodeView/state/ErgoState.scala

```diff
@@ -76,47 +77,55 @@ object ErgoState extends ScorexLogging {
+    * @param currentStateContext to be used for tx execution
+    * @param checkBoxExistence function to provide ErgoBox by BoxId
     * @return Result of transactions execution with total cost inside
     */
   def execTransactions(transactions: Seq[ErgoTransaction],
                        currentStateContext: ErgoStateContext)
                       (checkBoxExistence: ErgoBox.BoxId => Try[ErgoBox]): ValidationResult[Long] = {
-    import cats.implicits._
     val verifier: ErgoInterpreter = ErgoInterpreter(currentStateContext.currentParameters)
 
     @tailrec
-    def execTx(txs: List[ErgoTransaction], accCostTry: ValidationResult[Long]): ValidationResult[Long] = (txs, accCostTry) match {
-      case (tx :: tail, r: Valid[Long]) =>
-        val boxesToSpendTry: Try[List[ErgoBox]] = tx.inputs.toList
-          .map(in => checkBoxExistence(in.boxId))
-          .sequence
-
-        lazy val dataBoxesTry: Try[List[ErgoBox]] = tx.dataInputs.toList
-          .map(in => checkBoxExistence(in.boxId))
-          .sequence
-
-        lazy val boxes: Try[(List[ErgoBox], List[ErgoBox])] = dataBoxesTry.flatMap(db => boxesToSpendTry.map(bs => (db, bs)))
-
-        val vs = tx.validateStateless()
-          .validateNoFailure(txBoxesToSpend, boxesToSpendTry)
-          .validateNoFailure(txDataBoxes, dataBoxesTry)
-          .payload[Long](r.value)
-          .validateTry(boxes, e => ModifierValidator.fatal("Missed data boxes", e)) { case (_, (dataBoxes, toSpend)) =>
-            tx.validateStateful(toSpend.toIndexedSeq, dataBoxes.toIndexedSeq, currentStateContext, r.value)(verifier).result
-          }
-
-        execTx(tail, vs)
-      case _ =>
-        accCostTry
+    def collectBoxesById(
+                 remainingBoxIds: Iterator[ErgoBox.BoxId],
+                 resultingBoxes: Try[VectorBuilder[ErgoBox]] = Success(new VectorBuilder[ErgoBox])
+               ): Try[IndexedSeq[ErgoBox]] = {
+      if (!remainingBoxIds.hasNext) {
+        resultingBoxes.map(_.result())
+      } else {
+        checkBoxExistence(remainingBoxIds.next()) match {
+          case Success(box) =>
+            collectBoxesById(remainingBoxIds, resultingBoxes.map(_ += box))
+          case Failure(ex) =>
+            Failure(ex)
+        }
+      }
     }
 
     // Skip v1 block transactions validation if corresponding setting is on
     if (currentStateContext.blockVersion == 1 &&
           currentStateContext.ergoSettings.nodeSettings.skipV1TransactionsValidation) {
       Valid(0L)
     } else {
-      execTx(transactions.toList, Valid[Long](0L))
+      import spire.syntax.all.cfor
+      var costResult: ValidationResult[Long] = Valid[Long](0L)
+      cfor(0)(_ < transactions.length && costResult.isValid, _ + 1) { i =>
+        val validCostResult = costResult.asInstanceOf[Valid[Long]]
+        val tx = transactions(i)
+        val boxesToSpendTry: Try[IndexedSeq[ErgoBox]] = collectBoxesById(tx.inputs.iterator.map(_.boxId))
```

Comment:
Here you can take advantage of knowing val n = tx.inputs.length and save O(log(n)) allocations in builder by using sizeHint method.

Reply:
yes but `tx.inputs` and tx.dataInputs` have different types so `.map(_.boxId)` here was used for readability ...

Reply:
Maybe we should introduce new interface for `Input` and `DataInput` with `dev boxId: BoxId` ? We could then with any input that has `boxId` ... WYT ?

Reply:
collectBoxesById implicitly creates an empty builder here.  However, since we know the size, we can use sizeHint of the builder, to pre-allocate space.

Reply:
but `ArrayBuilder` does not take size hints, one can either `new mutable.ArrayBuilder.ofRef[ErgoBox]()` or `mutable.ArrayBuilder.make()`

Reply:
I see, strange that they did not make it part of the constructor